### PR TITLE
liveness: add health status liveness probe sidecar

### PR DESCRIPTION
**Pull request by pkalever**
# Describe what this PR does

The health status liveness probe shares and runs within the liveness-prometheus container. The health status liveness probe listen and serve requests at a dedicated port and path. By default they listen at '/healthz' path and '9680' port, which can be easily configurable.

Useful logs:
```
[0] pkalever 😎 ceph-csi✨ kubectl logs csi-rbdplugin-5dzrn liveness-prometheus                                   
I1104 20:16:06.923915  212212 cephcsi.go:125] Driver version: canary and Git version: 59aff136b51f38bdbca65bb0782215b7b786f89b
I1104 20:16:06.924212  212212 cephcsi.go:171] Starting driver type: liveness with name: liveness.csi.ceph.com     
I1104 20:16:06.924263  212212 liveness.go:122] Liveness Running                  
I1104 20:16:06.924458  212212 httpserver.go:21] Serving Metrics requests on: http://192.168.39.188:8680/metrics   
I1104 20:16:06.924555  212212 liveness.go:133] Serving Health requests on: http://192.168.39.188:9680/healthz
I1104 20:16:06.924739  212212 connection.go:153] Connecting to unix:///csi/csi.sock                               
I1104 20:16:06.927573  212212 liveness.go:110] CSI driver: "rbd.csi.ceph.com", Endpoint: unix:///csi/csi.sock
I1104 20:16:21.928003  212212 liveness.go:73] Metrics req: Sending probe request to CSI driver: "rbd.csi.ceph.com"
I1104 20:16:21.930418  212212 liveness.go:87] Metrics req: Health check succeeded
[...]
I1104 20:17:21.931447  212212 liveness.go:87] Metrics req: Health check succeeded
I1104 20:17:36.928015  212212 liveness.go:73] Metrics req: Sending probe request to CSI driver: "rbd.csi.ceph.com"
I1104 20:17:36.930653  212212 liveness.go:87] Metrics req: Health check succeeded
I1104 20:17:48.331070  212212 liveness.go:48] Healthz req: Sending probe request to CSI driver "rbd.csi.ceph.com" 
I1104 20:17:48.332878  212212 liveness.go:66] Healthz req: Health check succeeded

```

useful commands:
```
# kubectl exec -it  csi-rbdplugin-5dzrn -c csi-rbdplugin -- curl -X GET http://192.168.39.188:8680/metrics
# kubectl exec -it  csi-rbdplugin-5dzrn -c csi-rbdplugin -- curl -X GET http://192.168.39.188:9680/healthz

Also,
# minikube ssh
# docker ps | grep csi-rbdplugin-
# docker stop <contaner-ID> 

```

Fixes: #1096

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

### CODE REVIEW

File: cmd/cephcsi.go
Comment (nixpanic) [inline cmd/cephcsi.go:73]: ports are integers, not strings

File: deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml
Comment (nixpanic) [inline deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml:157]: Just because it "looks long" is not a good reason. Maybe you can look for other projects that do something like this and see what time they default to?

File: .commitlintrc.yml
Comment (nixpanic) [inline .commitlintrc.yml:44]: keep these alphabetically sorted please

File: cmd/cephcsi.go
Comment (pkalever) [reply cmd/cephcsi.go:69]: Technically yes, I'm trying to save the conversion from int to string here, as the JoinHostPort() from net package expects port in string form.

Also see:
https://github.com/kubernetes-csi/livenessprobe/blob/bba64df584a52d98ddf1904b7f8ec20a2828257c/cmd/livenessprobe/main.go#L39

File: internal/liveness/liveness.go
Comment (pkalever) [reply internal/liveness/liveness.go:40]: OK. I will look, how to make it better.

File: cmd/cephcsi.go
Comment (pkalever) [reply cmd/cephcsi.go:73]: Yeah, my previous comments should answer this.

File: .commitlintrc.yml
Comment (pkalever) [reply .commitlintrc.yml:44]: sure.

File: deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml
Comment (pkalever) [reply deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml:157]: Yes, I should have been clear before, got inspired by:
1. https://github.com/kubernetes-csi/livenessprobe/blob/bba64df584a52d98ddf1904b7f8ec20a2828257c/deployment/kubernetes/livenessprobe-sidecar.yaml#L21
2. https://github.com/kubernetes-csi/livenessprobe#usage
    https://github.com/kubernetes-csi/livenessprobe#command-line-options
   https://github.com/kubernetes-csi/livenessprobe/blob/master/deployment/kubernetes/livenessprobe-sidecar.yaml#L21
3. https://github.com/openshift/csi-livenessprobe#usage
   https://github.com/openshift/csi-livenessprobe#other-recognized-arguments
   https://github.com/openshift/csi-livenessprobe/blob/master/deployment/kubernetes/livenessprobe-sidecar.yaml#L21

**Comment by pkalever** [comment]
@nixpanic please take a look. Thanks!

**Comment by pkalever** [comment]
Not a related issue, says :
```
NAME            DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
csi-rbdplugin   1         1         0       1            0           <none>          20m
[Timeout] Failed to get daemonset
script returned exit code 1
```

File: cmd/cephcsi.go
Comment (Madhu-1) [reply cmd/cephcsi.go:69]: won't it cause an upgrade issue?  if someone just updated the image (which is the case for minor releases)

File: cmd/cephcsi.go
Comment (Madhu-1) [reply cmd/cephcsi.go:73]: why we need one more port? cant we just add one more endpoint to the metrics port?

File: deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml
Comment (Madhu-1) [inline deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml:157]: you can always pool the CSIDriver when you get a request to get live status.

File: internal/liveness/liveness.go
Comment (Madhu-1) [inline internal/liveness/liveness.go:59]: won't `err` value overwritten when you call w.Write()?

File: internal/liveness/liveness.go
Comment (Madhu-1) [inline internal/liveness/liveness.go:70]: looks like this can go to helper function

File: internal/liveness/liveness.go
Comment (Madhu-1) [inline internal/liveness/liveness.go:80]:
```suggestion
resp:="ok"
statuscode=http.StatusOK
ready, err := rpc.Probe(ctx, c.conn)
if err!=nil{
resp=err.Error()
statuscode=http.StatusInternalServerError
}else{
if !ready{
	resp:="Healthz req: driver responded but is not ready"
	statuscode=http.StatusInternalServerError
	}
}	
w.WriteHeader(statuscode)
_, err = w.Write([]byte(resp))
if err != nil {
	util.ErrorLog(ctx, "Healthz req: write failed: %v", err)
}
```

File: internal/liveness/liveness.go
Comment (Madhu-1) [inline internal/liveness/liveness.go:90]: ctx logging is not much helpful as it doesn't contain any information on what we need

File: internal/liveness/liveness.go
Comment (Madhu-1) [inline internal/liveness/liveness.go:152]: why we are starting one more server. is it needed?

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
I would suggest to keep the parameters very minimal and make use of available parameters

File: cmd/cephcsi.go
Comment (pkalever) [reply cmd/cephcsi.go:69]: I don't think it will cause any upgrade issues.

File: cmd/cephcsi.go
Comment (pkalever) [reply cmd/cephcsi.go:73]: OK, this is something I followed from the kube liveness probe project, its like giving a provision for users. I'm fine if we just want to use a single port for both metrics and health status.

Check: https://github.com/ceph/ceph-csi/pull/1560/files#diff-e3217918d2c8805d4f5446edf4350a9ee17a8616deb3ce92352a83765c730db5R163

File: deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml
Comment (pkalever) [reply deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml:157]: Maybe you mean probe? yes, we can always probe check.

As pointed other referenced projects are using 2sec and I feel 15 sec is a good time, but I will leave it up to the maintainers.

File: internal/liveness/liveness.go
Comment (pkalever) [reply internal/liveness/liveness.go:59]: good catch, will fix it.

File: internal/liveness/liveness.go
Comment (pkalever) [reply internal/liveness/liveness.go:70]: sure

File: internal/liveness/liveness.go
Comment (pkalever) [reply internal/liveness/liveness.go:90]: OK. Will, it hurt if we have ctx?

File: internal/liveness/liveness.go
Comment (pkalever) [reply internal/liveness/liveness.go:152]: Yeah, this is only needed if we want to listen on separate ports.
For a single port and multiple paths like: '/metrics' and '/healthz', we can live with a single server

**Review by pkalever — COMMENTED** [review]
@Madhu-1 Thanks for the review, will await for your opinion.

File: internal/liveness/liveness.go
Comment (Madhu-1) [reply internal/liveness/liveness.go:90]: it won't hurt, but as we have a separate function for logging without context let's use it.

File: internal/liveness/liveness.go
Comment (Madhu-1) [reply internal/liveness/liveness.go:152]: I would suggest to start one server not multiple servers

File: cmd/cephcsi.go
Comment (Madhu-1) [reply cmd/cephcsi.go:69]: can we please check once?

**Comment by Madhu-1** [comment]
one general suggestion would be to just add a new URL to the current liveness server  not to have one more server

**Pull request closed without merging**
